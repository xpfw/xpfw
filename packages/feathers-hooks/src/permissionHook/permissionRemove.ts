import { Method } from "../globals"
import { ValidateHook } from "../typeDef"
import generalPermissionHook from "./generalPermissionHook"

const permissionRemove: ValidateHook = (form, params) => {
  return generalPermissionHook(form, Method.Remove, params)
}

export default permissionRemove
