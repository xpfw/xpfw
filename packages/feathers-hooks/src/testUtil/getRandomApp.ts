import express from "@feathersjs/express"
import rest from "@feathersjs/express/rest"
import feathers from "@feathersjs/feathers"
import memdb from "feathers-memory"
import mongoService from "feathers-mongodb"
import { MongoClient } from "mongodb"

const memd: any = memdb
const getRandomApp = async (memoryServiceName: string, useMongo?: boolean) => {
  const f: any = feathers()
  const app: any = express.default(f)
  app.use(express.json())
  app.use(express.urlencoded({ extended: true }))
  let col: any | undefined
  let db: any | undefined
  let c: any
  app.configure(rest())
  if (useMongo) {
    c = await MongoClient.connect(`mongodb://localhost:27017/`)
    db  = c.db("feathersxpfwvalidatetests")
    col = db.collection(`testFor${memoryServiceName}`)
    app.use(memoryServiceName, mongoService({
      Model: col,
      paginate: {default: 10, max: 100}
    }))
  } else {
    app.use(memoryServiceName, memd({paginate: {default: 10, max: 100}}))
  }
  return Promise.resolve({
    app,
    service: app.service(memoryServiceName),
    cleanUpDb: async () => {
      if (useMongo) {
        await col.drop()
        await c.close()
      }
    }
  })
}

export default getRandomApp
