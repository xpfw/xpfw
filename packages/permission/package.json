{
  "name": "@xpfw/permission",
  "version": "1.0.1",
  "main": "dist",
  "typings": "dist",
  "license": "MIT",
  "peerDependencies": {
    "ajv": "^7.x.x",
    "lodash": "^4.x.x",
    "mobx": "^6.x.x",
    "react": "^17.x.x"
  },
  "scripts": {
    "build": "./node_modules/.bin/rollup -c",
    "dev": "./node_modules/.bin/rollup -c -w",
    "test": "./node_modules/.bin/jest --ci",
    "update": "ncu -a",
    "prepare": "npm run build"
  },
  "devDependencies": {
    "@types/ajv": "^1.0.0",
    "@types/jest": "^26.0.20",
    "@types/lodash": "^4.14.168",
    "@types/mockdate": "^2.0.0",
    "@types/react": "^17.0.2",
    "ajv": "^7.1.1",
    "jest": "^26.6.3",
    "jest-cli": "^26.6.3",
    "lodash": "^4.17.20",
    "mobx": "^6.1.6",
    "react": "^17.0.1",
    "rollup": "^2.39.0",
    "rollup-plugin-typescript2": "^0.29.0",
    "rollup-plugin-uglify": "^6.0.4",
    "ts-jest": "=26.5.1",
    "tslib": "^2.1.0",
    "typescript": "^4.1.5",
    "uglify-es": "^3.3.10"
  },
  "jest": {
    "globals": {
      "ts-jest": {
        "diagnostics": false
      }
    },
    "transform": {
      ".(ts|tsx)": "ts-jest"
    },
    "testPathIgnorePatterns": [
      "\\.snap$",
      "<rootDir>/node_modules/",
      "<rootDir>/lib",
      "<rootDir>/dist"
    ],
    "moduleNameMapper": {
    },
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js"
    ],
    "testURL": "http://localhost/"
  }
}
