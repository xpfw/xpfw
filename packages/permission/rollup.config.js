// rollup.config.js
import typescript from "rollup-plugin-typescript2"

export default {
  input: 'src/index.ts',
  output: {
    file: 'dist/index.js',
    format: 'cjs'
  },
  plugins: [
    typescript()
  ],
  external: [
    "ajv",
    "ajv-formats",
    "lodash",
    "mobx",
    "react",
    "react-dom"
  ]
};