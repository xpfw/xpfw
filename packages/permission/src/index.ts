import {
  GetPermissionCustom, GetPermissions, IParameters,
  IParametersBase, IParametersCustom, IPermissionSchema,
  IPermTypes, Permission
} from "./permissionSchema"
import validatePermission from "./validatePermission"

export {
  GetPermissionCustom, GetPermissions, IParameters,
  IParametersBase, IParametersCustom, IPermissionSchema,
  IPermTypes, Permission, validatePermission
}
