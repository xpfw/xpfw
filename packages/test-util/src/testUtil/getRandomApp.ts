import { authenticate, AuthenticationService, JWTStrategy } from "@feathersjs/authentication"
import { hooks, LocalStrategy } from "@feathersjs/authentication-local"
import * as express from "@feathersjs/express"
import * as rest from "@feathersjs/express/rest"
import feathers, { Hook, HookContext } from "@feathersjs/feathers"
import * as sios from "@feathersjs/socketio"
import * as memdb from "feathers-memory"
import * as mongoService from "feathers-mongodb"
import { get, isString } from "lodash"
import { MongoClient } from "mongodb"
import convertIds from "./convertIds"
import emptyPort from "./emptyPort"

const sio: any = sios
const res: any = rest
const mongoServic: any = mongoService
const memd: any = memdb
const promisifyListen = (app: any, port: number) => {
  return new Promise((resolve) => {
    let server: any
    server = app.listen(port, () => {
      resolve(server)
    })
  })
}

const channelName = "authenticated"

const getRandomApp = async (memoryServiceNames: string | string[],
                            useMongo?: boolean, ClientHolder?: any, useRest: boolean = false,
                            options?: {
                              userStore?: any, convertIds?: boolean, dbStore?: any, collections?: string[]
                            }) => {
  const f: any = feathers()

  const app: any = express.default(f)
  app.use(express.json())
  app.use(express.urlencoded({ extended: true }))
  let col: any | undefined
  let db: any | undefined
  let c: any
  app.configure(useRest ? res() : sio())
  app.set("authentication", {
    secret: "test_secret",
    entity: "user",
    service: "users",
    authStrategies: [ "jwt", "local" ],
    jwtOptions: {
      header: { typ: "access" },
      audience: "https://xpfw.github.io",
      issuer: "feathers",
      algorithm: "HS256",
      expiresIn: "1d"
    },
    local: {
      usernameField: "email", passwordField: "password"
    }
  })
  const authService = new AuthenticationService(app)
  authService.register("jwt", new JWTStrategy())
  authService.register("local", new LocalStrategy())
  app.use("/authentication", authService)
  let port = -1
  if (ClientHolder) {
    port = await emptyPort()
  }
  if (isString(memoryServiceNames)) {
    memoryServiceNames = [memoryServiceNames]
  }
  if (useMongo) {
    c = await MongoClient.connect(`mongodb://localhost:27017/`, { useNewUrlParser: true })
    db  = c.db("feathersxpfwvalidatetests" + port)
    for (const memoryServiceName of memoryServiceNames) {
      col = db.collection(`testFor${memoryServiceName}`)
      app.use(memoryServiceName, mongoServic({
        Model: col,
        paginate: {default: 10, max: 100},
        whitelist: ["$regex", "$options"]
      }))
      if (get(options, "convertIds", false)) {
        app.service(memoryServiceName).hooks({
          before: {
            create: [convertIds("_id", false)],
            update: [convertIds("_id", false)],
            patch: [convertIds("_id", false)],
            find: [convertIds("_id", true)]
          }
        })
      }
    }
  } else {
    for (const memoryServiceName of memoryServiceNames) {
      app.use(memoryServiceName, memd({paginate: {default: 10, max: 100}}))
    }
  }
  if (!useRest) {
    for (const memoryServiceName of memoryServiceNames) {
      app.service(memoryServiceName).publish((data: any, context: any) => {
        return app.channel(channelName)
      })
    }
  }
  app.use("users", memd({paginate: {default: 10, max: 100}}))
  app.service("users").hooks({
    // Make sure `password` never gets sent to the client
    after: hooks.protect("password"),
    before: {
      find: [
        authenticate("jwt")
      ],
      create: [
        hooks.hashPassword("password")
      ]
    }
  })

  app.on("login", (payload: any, { connection }: any) => {
    if (connection) {
      app.channel(channelName).join(connection)
    }
  })
  let server: any
  if (ClientHolder) {
    server = await promisifyListen(app, port)
    ClientHolder.connectTo(`http://localhost:${port}`, {
      useRest, userStore: get(options, "userStore"),
      dbStore: get(options, "dbStore"), collections: get(options, "collections")
    })
  }
  return Promise.resolve({
    app,
    port,
    cleanUp: async () => {
      if (useMongo) {
        await db.dropDatabase()
        await c.close()
      }
      if (ClientHolder) {
        ClientHolder.disconnect(useRest)
        server.close()
      }
    }
  })
}

export default getRandomApp
