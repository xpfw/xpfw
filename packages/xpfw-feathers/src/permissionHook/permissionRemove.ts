
import { globals } from "@xpfw/validate"
import { ValidateHook } from "../typeDef"
import generalPermissionHook from "./generalPermissionHook"

const permissionRemove: ValidateHook = (form, params) => {
  return generalPermissionHook(form, globals.Method.Remove, params)
}

export default permissionRemove
