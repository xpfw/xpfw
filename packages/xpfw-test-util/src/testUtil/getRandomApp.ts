import authentication from "@feathersjs/authentication"
import jwt from "@feathersjs/authentication-jwt"
import local, { hooks } from "@feathersjs/authentication-local"
import * as handler from "@feathersjs/errors/handler"
import * as express from "@feathersjs/express"
import * as rest from "@feathersjs/express/rest"
import feathers, { Application, Service } from "@feathersjs/feathers"
import * as sios from "@feathersjs/socketio"
import * as memdb from "feathers-memory"
import * as mongoService from "feathers-mongodb"
import { get, isString } from "lodash"
import { MongoClient } from "mongodb"
import convertIds from "./convertIds"
import emptyPort from "./emptyPort"

const sio: any = sios
const res: any = rest
const mongoServic: any = mongoService
const memd: any = memdb
const auth: any = authentication
const promisifyListen = (app: any, port: number) => {
  return new Promise((resolve) => {
    const server = app.listen(port, () => {
      resolve(server)
    })
  })
}

const channelName = "authenticated"

const getRandomApp = async (memoryServiceNames: string | string[],
                            useMongo?: boolean, ClientHolder?: any, useRest: boolean = false,
                            options?: {
                              userStore?: any, convertIds?: boolean, dbStore?: any, collections?: string[]
                            }) => {
  const app: any = express.default(feathers())
  app.use(express.json())
  app.use(express.urlencoded({ extended: true }))
  let col: any | undefined
  let db: any | undefined
  let c: any
  app.configure(useRest ? res() : sio())
  app.configure(auth({secret: "myawfulsecret"}))
  app.configure(local())
  app.configure(jwt())
  let port = -1
  if (ClientHolder) {
    port = await emptyPort()
  }
  if (isString(memoryServiceNames)) {
    memoryServiceNames = [memoryServiceNames]
  }
  if (useMongo) {
    c = await MongoClient.connect(`mongodb://localhost:27017/`)
    db  = c.db("feathersxpfwvalidatetests" + port)
    for (const memoryServiceName of memoryServiceNames) {
      col = db.collection(`testFor${memoryServiceName}`)
      app.use(memoryServiceName, mongoServic({
        Model: col,
        paginate: {default: 10, max: 100}
      }))
      if (get(options, "convertIds", false)) {
        app.service(memoryServiceName).hooks({
          before: {
            create: [convertIds("_id", false)],
            update: [convertIds("_id", false)],
            patch: [convertIds("_id", false)],
            find: [convertIds("_id", true)]
          }
        })
      }
    }
  } else {
    for (const memoryServiceName of memoryServiceNames) {
      app.use(memoryServiceName, memd({paginate: {default: 10, max: 100}}))
    }
  }
  if (!useRest) {
    for (const memoryServiceName of memoryServiceNames) {
      app.service(memoryServiceName).publish((data: any, context: any) => {
        return app.channel(channelName)
      })
    }
  }
  app.use("users", memd({paginate: {default: 10, max: 100}}))
  app.service("users").hooks({
    // Make sure `password` never gets sent to the client
    after: hooks.protect("password"),
    before: {
      find: [
        auth.hooks.authenticate("jwt")
      ],
      create: [
        hooks.hashPassword({ passwordField: "password" })
      ]
    }
  })
  app.service("authentication").hooks({
    before: {
     create: [
      // You can chain multiple strategies
      auth.hooks.authenticate(["jwt", "local"])
     ],
     remove: [
      auth.hooks.authenticate("jwt")
     ]
    }
   })

  app.on("login", (payload: any, { connection }: any) => {
    if (connection) {
      app.channel(channelName).join(connection)
    }
  })
  let server: any
  if (ClientHolder) {
    server = await promisifyListen(app, port)
    ClientHolder.connectTo(`http://localhost:${port}`, {
      useRest, userStore: get(options, "userStore"),
      dbStore: get(options, "dbStore"), collections: get(options, "collections")
    })
  }
  return Promise.resolve({
    app,
    port,
    cleanUp: async () => {
      if (useMongo) {
        await db.dropDatabase()
        await c.close()
      }
      if (ClientHolder) {
        ClientHolder.disconnect(useRest)
        server.close()
      }
    }
  })
}

export default getRandomApp
