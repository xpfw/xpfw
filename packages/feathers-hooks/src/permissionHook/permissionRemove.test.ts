import { Permission } from "@xpfw/permission"
import getRandomApp from "../testUtil/getRandomApp"
import permissionRemove from "./permissionRemove"

test("verify func", async () => {
  const serviceName = "permissionRemove"
  const { app, service } = await getRandomApp(serviceName)
  const removePermHook = permissionRemove({required: {create: Permission.Public, remove: Permission.Owner}}, {
    idPath: "id", directOwnershipPath: "belongsTo", serviceName
  })
  service.hooks({before: {remove: [ removePermHook ]}})

  const owningUserId = "asdf"
  const creationResultId = await service.create({missing: "required", belongsTo: owningUserId},
    {user: {_id: owningUserId}})
  expect(creationResultId).toMatchSnapshot("result of successfull creation")

  await expect(service.remove(creationResultId.id)).rejects.toMatchSnapshot("Forbidden remove")
  await expect(service.remove(creationResultId.id, {user: {id: owningUserId}}))
  .resolves.toMatchSnapshot("Allowed remove")
})
