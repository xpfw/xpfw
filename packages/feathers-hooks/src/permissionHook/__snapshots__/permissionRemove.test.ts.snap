// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verify func: Allowed remove 1`] = `
Object {
  "belongsTo": "asdf",
  "id": 0,
  "missing": "required",
}
`;

exports[`verify func: Forbidden remove 1`] = `[Forbidden: insufficient permission]`;

exports[`verify func: result of successfull creation 1`] = `
Object {
  "belongsTo": "asdf",
  "id": 0,
  "missing": "required",
}
`;
